{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import itertools\n",
    "from math import isclose\n",
    "import logging\n",
    "import os\n",
    "import shutil\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.rinterface_lib.callbacks import logger as rpy2_logger\n",
    "from rpy2.robjects import pandas2ri"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rpy2_logger.setLevel(logging.ERROR)\n",
    "base = importr('base')\n",
    "\n",
    "valid_yes_no = ['y', 'n']\n",
    "workdir_files = [f for f in glob.glob('working_directory' + '/*', recursive=True) if os.path.isfile(f)]\n",
    "workdir_files = [sub.replace('\\\\', '/') for sub in workdir_files]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_workdir_files():\n",
    "    if len(workdir_files) == 0:\n",
    "        raise ValueError('There are no valid files in the working directory mounted to this container.')\n",
    "    print('\\tAvailable files in working directory')\n",
    "    print('\\t--------------------------------')\n",
    "    for i, file in enumerate(workdir_files):\n",
    "        backslash_char = \"/\"\n",
    "        print(f\"\\t{i+1}: {file.split(backslash_char)[1]}\")\n",
    "\n",
    "def print_r_objects(rdata):\n",
    "    if len(rdata) == 0:\n",
    "        raise ValueError('There are no valid R objects loaded from this file.')\n",
    "    print('\\n')\n",
    "    print('\\tAvailable R objects')\n",
    "    print('\\t--------------------------------')\n",
    "    for i, o in enumerate(rdata):\n",
    "        print(f\"\\t{i+1}: {o}\")\n",
    "    print(\"\\n\")\n",
    "    r_obj_index = int(input(\"\\tWhich object is the expression data: \\t\").lower())\n",
    "    return rdata[r_obj_index - 1]\n",
    "\n",
    "def isfloat(num):\n",
    "    try:\n",
    "        float(num)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# question0_1\n",
    "- determines workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n\")\n",
    "valid_question0_1 = ['1', '2', '3']\n",
    "print(\"\\tAvailable tasks \")\n",
    "print(\"\\t-----------------------------------\")\n",
    "print(\"\\n\")\n",
    "print(\"\\t1) I want to download a pre-simulated spatial transcriptomics dataset.\")\n",
    "print(\"\\t2) I want to generate a parameter file using the interactive interface.\")\n",
    "print(\"\\t3) I have a parameter file and want to run a simulation.\")\n",
    "print('\\n')\n",
    "\n",
    "question0_1 = input(\"\\tSelect your task (1/2/3): \\t\").lower()\n",
    "if question0_1 not in valid_question0_1:\n",
    "    raise ValueError(f'Please enter 1, 2, or 3 to select a dataset. You entered {question0_1}') \n",
    "print(\"\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if question0_1 == '3':\n",
    "    #TODO: ask for existing parameter file\n",
    "    print_workdir_files()\n",
    "    print('\\n')\n",
    "    workflow3_userinput = input(f\"\\tWhich file are you using as parameter file (1-{len(workdir_files)}): \")\n",
    "    simulation_file_name = workdir_files[int(workflow3_userinput)-1]\n",
    "    print(f'\\n\\tStart simulation...\\n')\n",
    "    os.makedirs(f'working_directory/output_files', exist_ok=True)\n",
    "    os.system(f'Rscript ./scripts/ExampleData.R {simulation_file_name}')\n",
    "    print(f'\\n\\tSaved output file to your working directory/output_files')\n",
    "    sys.exit()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow1\n",
    "- Download pre-simulated data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if question0_1 == '1':\n",
    "    question1_1_filepaths_count = {\n",
    "        '1': 'example1_count.tsv',\n",
    "        '2': 'example2_count.tsv',\n",
    "        '3': 'example3_count.tsv'\n",
    "    }\n",
    "\n",
    "    question1_1_filepaths_meta = {\n",
    "        '1': 'example1_meta.tsv',\n",
    "        '2': 'example2_meta.tsv',\n",
    "        '3': 'example3_meta.tsv'\n",
    "    }\n",
    "\n",
    "    question1_1_filepaths_expr_pattern = {\n",
    "        '1': 'example1_expr_pattern.tsv',\n",
    "        '2': 'example2_expr_pattern.tsv',\n",
    "        '3': 'example3_expr_pattern.tsv'\n",
    "    }\n",
    "\n",
    "    question1_1_filepaths_docker_docu = {\n",
    "        '1': 'example1_docker_docu.txt',\n",
    "        '2': 'example2_docker_docu.txt',\n",
    "        '3': 'example3_docker_docu.txt'\n",
    "    }\n",
    "\n",
    "    question1_1_filepaths_parameter = {\n",
    "        '1': 'example1_parameter.tsv',\n",
    "        '2': 'example2_parameter.tsv',\n",
    "        '3': 'example3_parameter.tsv'\n",
    "    }\n",
    "\n",
    "\n",
    "    print(\"\\tAvailable data sets \")\n",
    "    print(\"\\t-----------------------------------\")\n",
    "    print(\"\\n\")\n",
    "    print(\"\\t1) Simulated data for 4,751 genes in 4,000 cells of 6 cell types in two regions on a unit square, using normal breast data profiled by snRNAseq as reference.\")\n",
    "    print(\"\\t2) Simulated data for 10,000 genes in 500 spots of 6 cell types, using mouse brain data profiled by SeqFISH+ as reference.\")\n",
    "    print(\"\\t3) Simulated data for 550 genes in 10,000 cells of 6 cell types, using human ovarian cancer data profiled by MERFISH as reference.\")\n",
    "    print('\\n')\n",
    "    question1_1 = input(\"\\tWhich dataset do you want to download (1/2/3)?  \\t\").lower()\n",
    "    if question1_1 not in question1_1_filepaths_count:\n",
    "        raise ValueError(f'Please enter 1, 2, or 3 to select a dataset. You entered {question1_1}') \n",
    "    print(\"\\n\")\n",
    "    os.makedirs(f'working_directory/example_data', exist_ok=True)\n",
    "    shutil.copyfile(f'example_data/{question1_1_filepaths_count[question1_1]}', f'working_directory/example_data/{question1_1_filepaths_count[question1_1]}')\n",
    "    shutil.copyfile(f'example_data/{question1_1_filepaths_meta[question1_1]}', f'working_directory/example_data/{question1_1_filepaths_meta[question1_1]}')\n",
    "    shutil.copyfile(f'example_data/{question1_1_filepaths_expr_pattern[question1_1]}', f'working_directory/example_data/{question1_1_filepaths_expr_pattern[question1_1]}')\n",
    "    shutil.copyfile(f'example_data/{question1_1_filepaths_docker_docu[question1_1]}', f'working_directory/example_data/{question1_1_filepaths_docker_docu[question1_1]}')\n",
    "    shutil.copyfile(f'example_data/{question1_1_filepaths_parameter[question1_1]}', f'working_directory/example_data/{question1_1_filepaths_parameter[question1_1]}')\n",
    "    print(f'\\n\\tSaved simulated data: {question1_1_filepaths_count[question1_1]}, {question1_1_filepaths_meta[question1_1]}, {question1_1_filepaths_expr_pattern[question1_1]}, {question1_1_filepaths_docker_docu[question1_1]} and {question1_1_filepaths_parameter[question1_1]} to working_directory/example_data.')\n",
    "\n",
    "    sys.exit()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow2\n",
    "- Generate a simulation parameter file."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_1\n",
    "- Input expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select or read in expression data file\n",
    "parameters = {\n",
    "    'path_to_input_dir': 'working_directory/'\n",
    "}\n",
    "\n",
    "# QUESTION2_1\n",
    "question2_1_filepaths_expression = {\n",
    "    '1': 'InputData/expression_data/fake1_expr.Rdata',\n",
    "    '2': 'InputData/expression_data/fake2_expr.Rdata',\n",
    "    '3': 'InputData/expression_data/fake3_expr.Rdata',\n",
    "    '4': 'InputData/expression_data/snRNAseq_breast_expr_033023.RData',\n",
    "    '5': 'InputData/expression_data/SeqFishPlusCortex_expr_033023.Rdata',\n",
    "    '6': 'InputData/expression_data/MERFISH_OV_expr.RData',\n",
    "    '7': 'user_input'\n",
    "}\n",
    "\n",
    "question2_1_filepaths_cellfeature = {\n",
    "    '1': 'InputData/cell_feature_data/fake1_cellfeature.Rdata',\n",
    "    '2': 'InputData/cell_feature_data/fake2_cellfeature.Rdata',\n",
    "    '3': 'InputData/cell_feature_data/fake3_cellfeature.Rdata',\n",
    "    '4': 'InputData/cell_feature_data/snRNAseq_breast_cellfeature_033023.RData',\n",
    "    '5': 'InputData/cell_feature_data/SeqFishPlusCortex_cellfeature_033023.Rdata',\n",
    "    '6': 'InputData/cell_feature_data/MERFISH_OV_cellfeature.RData',\n",
    "    '7': 'user_input'\n",
    "}\n",
    "\n",
    "print(\"\\tWhat data do you want to use for simulation?\")\n",
    "print(\"\\t-----------------------------------\")\n",
    "print(\"\\n\")\n",
    "print(\"\\t1) Decoy data 1\")\n",
    "print(\"\\tDetails: It includes (1) count matrix for 10 genes by 1000 cells of 2 cell types, and (2) cell feature matrix for annotated cell type.\")\n",
    "print(\"\\t2) Decoy data 2\")\n",
    "print(\"\\tDetails: It includes (1) count matrix for 10 genes by 1000 cells of 2 cell types, and (2) cell feature matrix for annotated cell type and spatial coordinate.\")\n",
    "print(\"\\t3) Decoy data 3\")\n",
    "print(\"\\tDetails: It includes (1) count matrix for 10 genes by 1000 cells of 2 cell types, and (2) cell feature matrix for annotated cell type, spatial coordinate, and region.\")\n",
    "print(\"\\t4) Normal human breast snRNAseq data\")\n",
    "print(\"\\tIt includes (1) count matrix for 4751 genes by 5990 cells of 6 cell types (epithelial cell, adipocyte, fibroblast, endothelial cell, immune (myeloid) and muscle), and (2) cell feature matrix for annotated cell type. PMID: 35549429\")\n",
    "print(\"\\t5) Normal mouse brain SeqFISH+ data\")\n",
    "print(\"\\tDetails: It includes (1) count matrix for 10,000 genes by 511 cells of 6 cell types (excitatory neuron, interneuron, astrocyte, microglia, oligodendrocyte and endothelial cells), and (2)cell feature matrix including cell type annotation and spatial coordinate on 2D (x, y). PMID: 35549429\")\n",
    "print(\"\\t6) Ovarian cancer MERFISH data\")\n",
    "print(\"\\tDetails: It includes (1) count matrix for 550 genes by 355,633 cells of 6 cell types (tumor, adipose, endo-immune, macro-immune, macrophage, and others), and (2) cell feature matrix including cell type annotation and spatial coordinate on 2D. Data source: vizgen\")\n",
    "print(\"\\t7) User input\")\n",
    "print('\\n')\n",
    "question2_1 = input(\"\\tWhat data do you want to use for simulation? \")\n",
    "print('\\n')\n",
    "if question2_1 not in question2_1_filepaths_expression:\n",
    "    # this probably should ultimately be a while loop\n",
    "    raise ValueError(f'Please enter 1, 2, 3, 4, 5, 6 or 7 to select an expression data set. You entered {question2_1}') \n",
    "\n",
    "if question2_1_filepaths_expression[question2_1] != 'user_input':\n",
    "    question2_1_download = input(\"\\tWould you like to download this data table locally (y/n)? [default = y; select n if data is already downloaded for simulation]\").lower()\n",
    "    \n",
    "    if question2_1_download == '':\n",
    "        question2_1_download = 'y'\n",
    "        print(f\"\\tUsing default: {question2_1_download}\")\n",
    "    \n",
    "    if question2_1_download in valid_yes_no:\n",
    "        if question2_1_download == 'y':\n",
    "            #TODO: update with real data\n",
    "            expression_filepath_slim = question2_1_filepaths_expression[question2_1].split('/')[-1]\n",
    "            os.makedirs(f'working_directory/InputData', exist_ok=True)\n",
    "            os.makedirs(f'working_directory/InputData/expression_data', exist_ok=True)\n",
    "            shutil.copyfile(f'{question2_1_filepaths_expression[question2_1]}', f'working_directory/{question2_1_filepaths_expression[question2_1]}')\n",
    "            #parameters['expression_data_file'] = expression_filepath_slim\n",
    "            print(f'\\n\\tSaved expression data: {expression_filepath_slim} to your working directory.')\n",
    "            cellfeature_filepath_slim = question2_1_filepaths_cellfeature[question2_1].split('/')[-1]\n",
    "            os.makedirs(f'working_directory/InputData/cell_feature_data', exist_ok=True)\n",
    "            shutil.copyfile(f'{question2_1_filepaths_cellfeature[question2_1]}', f'working_directory/{question2_1_filepaths_cellfeature[question2_1]}')\n",
    "            #parameters['cell_feature_data_file'] = cellfeature_filepath_slim\n",
    "            print(f'\\n\\tSaved cell feature data: {cellfeature_filepath_slim} to your working directory.')\n",
    "            \n",
    "else:\n",
    "    print_workdir_files()\n",
    "    print('\\n')\n",
    "    question2_1_userinput_expression = input(f\"\\tUpload a G gene by N cell matrix for expression count data. Instruction: Row names should be unique identifiers of genes (1-{len(workdir_files)}): \")\n",
    "    question2_1_filepaths_expression['7'] = workdir_files[int(question2_1_userinput_expression)-1]\n",
    "    question2_1_userinput_cellfeature = input(f\"\\tUpload a N by K matrix for cell feature data. Instruction: Column and row names are not expected. Cells should be in the same order as the uploaded expression count data. The first column is required, which should be cell type annotation. Other columns are optional.  If spatial coordinates on x, y axes are provided, they should be the 2-3 columns of the input cell feature data. If a spatial region variable is provided, it should be the 4th column of the data. (1-{len(workdir_files)}): \")\n",
    "    question2_1_filepaths_cellfeature['7'] = workdir_files[int(question2_1_userinput_cellfeature)-1]\n",
    "    \n",
    "parameters['expression_data_file'] = question2_1_filepaths_expression[question2_1]\n",
    "parameters['cell_feature_data_file'] = question2_1_filepaths_cellfeature[question2_1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question2_1 continued\n",
    "# Identify expression data file type\n",
    "\n",
    "file_type_dict = {\n",
    "        '.rdata': 'Rdata',\n",
    "        '.tsv': '\\t',\n",
    "        '.csv': ','\n",
    "    }\n",
    "expression_data_file_extension = os.path.splitext(parameters['expression_data_file'])[-1].lower()\n",
    "if expression_data_file_extension not in file_type_dict:\n",
    "    raise ValueError('Please enter an .Rdata, .tsv, or .csv file for expression data.')\n",
    "\n",
    "parameters['expression_data_file_type'] = file_type_dict[expression_data_file_extension]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question2_1 continued\n",
    "# Read expression data and identify cell types\n",
    "pandas2ri.activate()\n",
    "if parameters['expression_data_file_type'] == 'Rdata': \n",
    "    loaded_r_obj = robjects.r.load(f\"{parameters['cell_feature_data_file']}\")\n",
    "    cellfeature_data_r = robjects.r.get(robjects.r.ls()[robjects.r.ls() != \"fileName\"])\n",
    "    cellfeature_data_r = robjects.conversion.rpy2py(cellfeature_data_r)\n",
    "    cellfeature_data_r_names = cellfeature_data_r.iloc[:,0]\n",
    "    cellfeature_data_cell_types = list(set(cellfeature_data_r_names))\n",
    "# TODO: add sample input\n",
    "else:\n",
    "    expression_data_cell_types = pd.read_csv(f\"{parameters['cell_feature_data_file']}\", sep=parameters['expression_data_file_type'], index_col=0).columns\n",
    "\n",
    "print('\\n')\n",
    "print(f'\\tUnique cell types detected in expression data: {len(cellfeature_data_cell_types)}')\n",
    "cellfeature_data_cell_types.sort()\n",
    "parameters['expression_data_cell_types'] = ','.join(cellfeature_data_cell_types)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_2 (deleted)\n",
    "- Input spatial data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question2_4\n",
    "- Simulate spatial map using parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(cellfeature_data_r.columns) == 1:\n",
    "    parameters['simulate_spatial_data'] = 'TRUE'\n",
    "    # Simulate spatial map using parameters.\n",
    "    print(\"\\n\")\n",
    "    question2_4_1 = input(\"\\tNumber of simulated cells [default =10,000]: \\t\").lower()\n",
    "    \n",
    "    if question2_4_1 == '':\n",
    "        question2_4_1 = '10000'\n",
    "        print(f\"\\tUsing default: {question2_4_1}\")\n",
    "\n",
    "    if not question2_4_1.isnumeric():\n",
    "        raise ValueError(f'Please enter an integer for the number of simulated cells. You entered {question2_4_1}')\n",
    "    \n",
    "    parameters['num_simulated_cells'] = question2_4_1\n",
    "\n",
    "    print(\"\\n\")\n",
    "    question2_4_2 = input(\"\\tNumber of regions [default = 2; suggested: 1-10]: \\t\").lower()\n",
    "    \n",
    "    if question2_4_2 == '':\n",
    "        question2_4_2 = '2'\n",
    "        print(f\"\\tUsing default: {question2_4_2}\")\n",
    "\n",
    "    if not question2_4_2.isnumeric():\n",
    "        raise ValueError(f'Please enter an integer for the number of regions. You entered {question2_4_2}')\n",
    "\n",
    "    parameters['num_regions'] = question2_4_2\n",
    "\n",
    "    print(\"\\n\")\n",
    "    question2_4_3a = input(\"\\tInput custom cell-type proportions in each region (y/n)? [default: n = all equals to cell-type proportions of the input expression data] \\t\").lower()\n",
    "    \n",
    "    if question2_4_3a == 'n' or question2_4_3a == '':\n",
    "        print('\\tUsing default cell-type proportions.')\n",
    "        print('\\n')\n",
    "        parameters['custom_cell_type_proportions'] = 'FALSE'\n",
    "        # ADD cell_type_proportion to parameter file\n",
    "        num_regions = int(question2_4_2)\n",
    "        cell_freq = dict((x,round(list(cellfeature_data_r_names).count(x)/cellfeature_data_r.shape[0], 2)) for x in cellfeature_data_cell_types)\n",
    "        n = 1\n",
    "        for i in range(1, num_regions + 1):\n",
    "            for cell_type in cellfeature_data_cell_types:\n",
    "                parameters[f'cell_type_proportion_{n}'] = f'{i},{cell_type},{cell_freq[cell_type]}'\n",
    "                n += 1\n",
    "   \n",
    "    else:\n",
    "        parameters['custom_cell_type_proportions'] = 'TRUE'\n",
    "        question2_4_3 = []\n",
    "\n",
    "        expression_data_cell_types_proportions = []\n",
    "        cell_types = list(set([t.split('.')[0] for t in cellfeature_data_cell_types]))\n",
    "        print('\\n')\n",
    "        print('\\tCell types detected: ', len(cell_types))\n",
    "    \n",
    "        p_index = 0\n",
    "        for r in range(0, int(parameters['num_regions'])):\n",
    "            region = r+1\n",
    "            res = {\n",
    "                'region': region,\n",
    "                'proportions': [],\n",
    "            }\n",
    "            print('\\n')\n",
    "            print(f'\\tRegion {region} (Must sum up to 1)')\n",
    "            print('\\t--------------')\n",
    "            for cell_type in cell_types:\n",
    "                cell_type_proportion = input(f'\\tProportion of cell type \"{cell_type}\" in region {region}: ')\n",
    "                if cell_type_proportion == '':\n",
    "                    cell_type_proportion = 0\n",
    "                else:\n",
    "                    cell_type_proportion = float(cell_type_proportion)\n",
    "                res['proportions'].append(cell_type_proportion)\n",
    "                parameters[f'cell_type_proportion_{p_index}'] = f'{region},{cell_type},{cell_type_proportion}'\n",
    "                p_index += 1\n",
    "            total_proportions = sum(res['proportions'])\n",
    "            print('\\n')\n",
    "            print(f'\\tSum of cell type proportions in region {region}: {total_proportions}')\n",
    "            expression_data_cell_types_proportions.append(res)\n",
    "            if not isclose(total_proportions, 1):\n",
    "                raise ValueError('Cell type proportions for a given region must sum up to 1.')\n",
    "\n",
    "    question2_4_4 = input(\"\\tSpecify cell-cell location interaction (inhibition/attraction) (y/n)? [default = n]\\nInstruction: Users can select cell type pairs and determine the strength. Strength <0 indicates cell-cell inhibition; and strength > 0 indicates cell-cell attraction.\\t\").lower()\n",
    "    \n",
    "    if question2_4_4 == '':\n",
    "        print('\\tUsing default n.')\n",
    "        print('\\n')\n",
    "        question2_4_4 = 'n'\n",
    "\n",
    "    if question2_4_4 not in valid_yes_no:\n",
    "        raise ValueError(\"Please enter y/n.\")\n",
    "\n",
    "    if question2_4_4 == 'n':\n",
    "        parameters['custom_cell_location_interactions'] = 'FALSE'\n",
    "    else:\n",
    "        parameters['custom_cell_location_interactions'] = 'TRUE'\n",
    "        question2_4_4 = ''\n",
    "        if_continue = 1\n",
    "        while if_continue:\n",
    "            print('\\n')\n",
    "            print(f'\\tCell type 1: Choose from cell types')\n",
    "            print('\\t--------------')\n",
    "            for i, cell_type in enumerate(cellfeature_data_cell_types):\n",
    "                print(f'\\t{i+1}: {cell_type}')\n",
    "            question2_4_4_cell1 = int(input(f\"\\tSelect cell type:\\t\"))\n",
    "            print(f'\\tCell type 2: Choose from cell types')\n",
    "            print('\\t--------------')\n",
    "            for i, cell_type in enumerate(cellfeature_data_cell_types):\n",
    "                print(f'\\t{i+1}: {cell_type}')\n",
    "            question2_4_4_cell2 = int(input(f\"\\tSelect cell type:\\t\"))\n",
    "            question2_4_4_strength = input(\"\\tValue of strength: (suggested value -2 to 2)\\t\")\n",
    "            question2_4_4 = f'{question2_4_4}{cellfeature_data_cell_types[question2_4_4_cell1 -1]},{cellfeature_data_cell_types[question2_4_4_cell2 - 1]},{question2_4_4_strength};'\n",
    "            question2_4_4_continue = input(\"\\tDo you want to continue specifying other pairs of cell-cell location interaction (y/n) [default = n]\").lower()\n",
    "            if question2_4_4_continue == '':\n",
    "                print('\\tUsing defulat n.')\n",
    "                print('\\n')\n",
    "                question2_4_4_continue = 'n'\n",
    "            if question2_4_4_continue == 'y':\n",
    "                if_continue = 1\n",
    "            else:\n",
    "                if_continue = 0\n",
    "        parameters['custom_cell_location_interactions'] = question2_4_4\n",
    "    \n",
    "    question2_4_5 = input(\"\\tThis parameter adjusts cells to be evenly distributed on a slide (range: 0-1). [default = 0]\\t\")\n",
    "    \n",
    "    if question2_4_5 == '':\n",
    "        print('\\tUsing default 0.')\n",
    "        print('\\n')\n",
    "        question2_4_5 = '0'\n",
    "\n",
    "    if not isfloat(question2_4_5):\n",
    "        raise ValueError(f'Please enter a numeric value.')\n",
    "    \n",
    "    if float(question2_4_5) < 0 or float(question2_4_5) > 1:\n",
    "        raise ValueError(f'Please enter a value between 0 and 1.')\n",
    "\n",
    "    parameters['cell_even_distribution'] = question2_4_5\n",
    "\n",
    "    question2_4_6 = input(\"\\tCells closer than this cutoff will be considered overlapping to each other and all but one will be removed (range: 0-0.02 of the slide length/width) [default = 0]:\\t\")\n",
    "    \n",
    "    if question2_4_6 == '':\n",
    "        print('\\tUsing default 0.')\n",
    "        print('\\n')\n",
    "        question2_4_6 = '0'\n",
    "\n",
    "    if not isfloat(question2_4_6):\n",
    "        raise ValueError(f'Please enter a numeric value.')\n",
    "\n",
    "    if float(question2_4_6) < 0 or float(question2_4_6) > 0.1:\n",
    "        raise ValueError('Please enter a value between 0 and 0.1')\n",
    "        \n",
    "    parameters['cell_overlap_cutoff'] = question2_4_6"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(cellfeature_data_r.columns) > 1:\n",
    "    print('\\n')\n",
    "    question2_5 = input(\"\\tDo you want to simulate new cells (y/n)? [default: n = do not simulate new cells but simulate new expression data for existing cells].\\t\").lower()\n",
    "\n",
    "    if question2_5 == '':\n",
    "        question2_5 = 'n'\n",
    "        print('\\n\\tUsing default: n')\n",
    "\n",
    "    if question2_5 not in valid_yes_no:\n",
    "        raise ValueError(\"Please enter y/n.\")\n",
    "    \n",
    "    if question2_5 == 'n':\n",
    "        parameters['simulate_spatial_data'] = 'FALSE'\n",
    "    \n",
    "    if question2_5 == 'y':\n",
    "        parameters['simulate_spatial_data'] = 'TRUE'\n",
    "        print('\\n')\n",
    "        question2_5_1 = input(\"\\tNumber of simulated cells [default = 10,000]: \\t\").lower()\n",
    "        \n",
    "        if question2_5_1 == '':\n",
    "            question2_5_1 = '10000'\n",
    "            print(f\"\\tUsing default: {question2_5_1}\")\n",
    "\n",
    "        if not question2_5_1.isnumeric():\n",
    "            raise ValueError(f'Please enter an integer for the number of simulated cells. You entered {question2_4_1}')\n",
    "        \n",
    "        parameters['num_simulated_cells'] = question2_5_1\n",
    "\n",
    "        print('\\n')\n",
    "        question2_5_3 = input(\"\\tCells closer than this cutoff will be considered overlapping to each other and all but one will be removed (range: 0-0.02 of the slide length/width) [default = 0]:\\t\")\n",
    "    \n",
    "        if question2_5_3 == '':\n",
    "            print('\\tUsing default 0')\n",
    "            print('\\n')\n",
    "            question2_5_3 = '0'\n",
    "\n",
    "        if not isfloat(question2_5_3):\n",
    "            raise ValueError('Please enter a value between 0 and 0.1')\n",
    "            \n",
    "        parameters['cell_overlap_cutoff'] = question2_5_3\n",
    "\n",
    "        print(\"\\n\")\n",
    "        print(\"\\tAvailable methods for determining window on existing ST data\")\n",
    "        print(\"\\t-----------------------------------\")\n",
    "        print(\"\\n\")\n",
    "        print(\"\\t1) rectangle\")\n",
    "        print(\"\\t2) convex\")\n",
    "        print(\"\\t3) convex2\")\n",
    "        print(\"\\t4) convex3\")\n",
    "        print(\"\\t5) convex5\")\n",
    "        print(\"\\t6) network\")\n",
    "        print('\\n')\n",
    "\n",
    "        question2_5_2 = input(\"\\tMethod for determining window on existing ST data [default = network]: \\t\").lower()\n",
    "        valid_question2_5_2 = {\n",
    "            \"1\": \"rectangle\", \n",
    "            \"2\": \"convex\", \n",
    "            \"3\": \"convex2\",\n",
    "            \"4\": \"convex3\",\n",
    "            \"5\": \"convex5\",\n",
    "            \"6\": \"network\"\n",
    "        }\n",
    "        \n",
    "        if question2_5_2 == '':\n",
    "            question2_5_2 = '6'\n",
    "            print(f\"\\tUsing default: {valid_question2_5_2[question2_5_2]}\")\n",
    "\n",
    "        if not question2_5_2.isnumeric():\n",
    "            raise ValueError(f'Please enter an integer for the number of simulated cells. You entered {question2_4_1}')\n",
    "        \n",
    "        parameters['window_method'] = valid_question2_5_2[question2_5_2]\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n\")\n",
    "print('\\t**** Provide simulation parameters for expression profiles ****')\n",
    "print(\"\\n\")\n",
    "question2_6_1 = input(\"\\tSequencing depth ratio between simulated and reference data [default = 1]? \\t\").lower()\n",
    "  \n",
    "if question2_6_1 == '':\n",
    "        question2_6_1 = '1'\n",
    "        print('\\n\\tUsing default: 1')\n",
    "\n",
    "if not isfloat(question2_6_1):\n",
    "    raise ValueError(f'Please enter a numeric value.')\n",
    "\n",
    "parameters['expr_depth_ratio'] = question2_6_1\n",
    "\n",
    "if len(cellfeature_data_r.columns) == 4 and len(list(set(cellfeature_data_r.iloc[:,3]))) > 1:\n",
    "    print(\"\\n\")\n",
    "    question2_6_7 = input(\"\\tDo you want to model the input expression data separately for each region? (y/n) [default : n]\\t\").lower()\n",
    "    if question2_6_7 == '':\n",
    "        print(\"\\t Using default n\")\n",
    "        print(\"\\n\")\n",
    "        question2_6_7 = 'n'\n",
    "    if question2_6_7 not in valid_yes_no:\n",
    "        raise ValueError('Please enter y/n.')\n",
    "    if question2_6_7 == 'y':\n",
    "        parameters['region_specific_model'] = 'TRUE'\n",
    "    else:\n",
    "        parameters['region_specific_model'] = 'FALSE'\n",
    "    cellfeature_num_regions = len(list(set(cellfeature_data_r.iloc[:,3])))\n",
    "else:\n",
    "    parameters['region_specific_model'] = 'NULL'\n",
    "    cellfeature_num_regions = 1\n",
    "    \n",
    "print(\"\\n\")\n",
    "question2_6_2 = input(\"\\tDo you want to mimic gene-gene correlation of the reference data (y/n)? [default: n, simulates independent genes]\\nNote: Select “y” only if the gene-gene correlation is pre-estimated for each cell type. STsimulator provides functions/pipelines for estimating and saving gene-gene correlation, which can be served as input here.\\t\").lower()\n",
    "\n",
    "if question2_6_2 == '':\n",
    "    question2_6_2 = 'n'\n",
    "    print('\\n\\tUsing default: ', question2_6_2)\n",
    "\n",
    "if question2_6_2 not in valid_yes_no:\n",
    "    raise ValueError('Please enter y/n.')\n",
    "\n",
    "if question2_6_2 == 'n':\n",
    "    parameters['gene_cor'] = 'FALSE'\n",
    "else:\n",
    "    parameters['gene_cor'] = 'TRUE'\n",
    "    print('\\n')\n",
    "    print_workdir_files()\n",
    "    question2_6_3 = input(\"\\tProvide a path for uploading a pre-estimated gene-gene correlation file\\t\").lower()\n",
    "    if question2_6_3 == '':\n",
    "        raise ValueError('Please select a file.') \n",
    "    elif not question2_6_3.isnumeric():\n",
    "        raise ValueError('Please select a file.')\n",
    "    else:\n",
    "        parameters['copula_input'] = workdir_files[int(question2_6_3)-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'question2_4_2' in globals():\n",
    "    sim_num_regions = int(question2_4_2)\n",
    "else:\n",
    "    sim_num_regions = 0\n",
    "\n",
    "if cellfeature_num_regions > 1 or sim_num_regions > 1:\n",
    "    \n",
    "    print(\"\\n\")\n",
    "    question2_6_4 = input(\"\\tAdd spatial patterns? (y/n) [default: n]\\t\").lower()\n",
    "\n",
    "    if question2_6_4 == 'y':\n",
    "        spatial_patterns = []\n",
    "        finished_question2_6_4 = False\n",
    "        while not finished_question2_6_4:\n",
    "            spatial_pattern = {}\n",
    "            print('\\n')\n",
    "            print('\\tPrompts for adding spatial pattern: ')\n",
    "            \n",
    "            if sim_num_regions > 1:\n",
    "                question2_6_4a = input(f\"\\tWhich region? [1-{sim_num_regions}]\\t\").lower()\n",
    "                if not question2_6_4a.isnumeric():\n",
    "                    raise ValueError(f\"Please specify a region that is an integer between 1 and {sim_num_regions}\")\n",
    "                spatial_pattern['region'] = question2_6_4a\n",
    "            \n",
    "            if cellfeature_num_regions > 1:\n",
    "                print('\\n')\n",
    "                print(f'\\tChoose from regions')\n",
    "                print('\\t--------------')\n",
    "                region_list = list(set(cellfeature_data_r.iloc[:,3]))\n",
    "                for i, region in enumerate(region_list):\n",
    "                    print(f'\\t{i+1}: {region}')\n",
    "                question2_6_5a = int(input(f'\\tPlease specify a region:\\t'))\n",
    "                spatial_pattern['region'] = region_list[question2_6_5a - 1]\n",
    "            \n",
    "            print('\\n')\n",
    "            print(f'\\tChoose from cell types')\n",
    "            print('\\t--------------')\n",
    "            cell_types = cellfeature_data_cell_types\n",
    "            for i, cell_type in enumerate(cell_types):\n",
    "                print(f'\\t{i+1}: {cell_type}')\n",
    "\n",
    "            question2_6_4b = int(input(f\"\\tSelect cell type:\\t\"))\n",
    "            spatial_pattern['cell_type'] = cell_types[question2_6_4b - 1]\n",
    "\n",
    "            question2_6_4c = input(f\"\\tDo you want to upload gene ID (y/n) [default: n. The proportion of genes with the pattern should be specified, and genes will be randomly selected.]\\t\").lower()\n",
    "            if question2_6_4c == '':\n",
    "                question2_6_4c = 'n'\n",
    "                print('\\tUsing default: n.')\n",
    "            \n",
    "            if question2_6_4c == 'y':\n",
    "                print_workdir_files()\n",
    "                print('\\n')\n",
    "                question2_6_4c_userinput = input(f\"\\tWhich file are you using as gene ID (1-{len(workdir_files)}): \")\n",
    "                geneid = pd.read_csv(workdir_files[int(question2_6_4c_userinput)-1], sep=\"\\t\", header=None)\n",
    "                spatial_pattern['gene_id'] = ','.join(list(geneid[0]))\n",
    "                spatial_pattern['gene_prop'] = 'NULL'\n",
    "            else:\n",
    "                spatial_pattern['gene_id'] = \"NULL\"\n",
    "                question2_6_4d = input(f\"\\tGene proportion (default = 0.1):\\t\")\n",
    "                if question2_6_4d == '':\n",
    "                    question2_6_4d = '0.1'\n",
    "                    print('\\tUsing default 0.1.')\n",
    "                spatial_pattern['gene_prop'] = question2_6_4d\n",
    "\n",
    "            question2_6_4e = input(f\"\\tMean effect at log(count) scale (default = 0.5):\")\n",
    "\n",
    "            if question2_6_4e == '':\n",
    "                question2_6_4e = '0.5'\n",
    "                print('\\tUsing default 0.5.')\n",
    "        \n",
    "            spatial_pattern['mean'] = question2_6_4e\n",
    "\n",
    "            question2_6_4f = input(f\"\\tSD of effect at log(count) scale (default = 0):\\t\")\n",
    "\n",
    "            if question2_6_4f == '':\n",
    "                question2_6_4f = '0'\n",
    "                print('\\tUsing default 0.')\n",
    "            spatial_pattern['sd'] = question2_6_4f\n",
    "\n",
    "            spatial_patterns.append(spatial_pattern)\n",
    "            print('\\n')\n",
    "            question2_6_4exit = input(f\"\\tSpecify another spatial pattern? (y/n) [default: n]\\t\").lower()\n",
    "            if question2_6_4exit == 'n' or question2_6_4exit == '':\n",
    "                finished_question2_6_4 = True\n",
    "\n",
    "        for i, spatial_patten in enumerate(spatial_patterns):\n",
    "            for k in spatial_patten.keys():\n",
    "                parameters[f'spatial_pattern_{i+1}_{k}'] = spatial_patten[k]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('\\n')\n",
    "question2_6_5 = input(\"\\tAdd cell-cell interactions - expression associated with cell-cell distance? (y/n) [default: n]\\t\").lower()\n",
    "\n",
    "if question2_6_5 == 'y':\n",
    "    cell_cell_interactions = []\n",
    "    finished_question2_6_5 = False\n",
    "    while not finished_question2_6_5:\n",
    "        cell_cell_interaction = {}\n",
    "        print('\\n')\n",
    "        print('\\tPrompts for adding cell-cell interaction: ')\n",
    "        if sim_num_regions > 1:\n",
    "            question2_6_5a = input(f\"\\tWhich region? [1-{sim_num_regions}]\\t\").lower()\n",
    "            if not question2_6_5a.isnumeric():\n",
    "                raise ValueError(f\"Please specify a region that is an integer between 1 and {sim_num_regions}\")\n",
    "            cell_cell_interaction['region'] = question2_6_5a\n",
    "        elif cellfeature_num_regions > 1:\n",
    "            print('\\n')\n",
    "            print(f'\\tChoose from regions')\n",
    "            print('\\t--------------')\n",
    "            region_list = list(set(cellfeature_data_r.iloc[:,3]))\n",
    "            for i, region in enumerate(region_list):\n",
    "                print(f'\\t{i+1}: {region}')\n",
    "            question2_6_5a = int(input(f'\\tPlease specify a region:\\t'))\n",
    "            cell_cell_interaction['region'] = region_list[question2_6_5a - 1]\n",
    "        else:\n",
    "            cell_cell_interaction['region'] = 'NULL'\n",
    "        \n",
    "        print('\\n')\n",
    "        print(f'\\tChoose from cell types')\n",
    "        print('\\t--------------')\n",
    "        cell_types = cellfeature_data_cell_types\n",
    "        for i, cell_type in enumerate(cell_types):\n",
    "            print(f'\\t{i+1}: {cell_type}')\n",
    "\n",
    "        question2_6_5b = int(input(f\"\\tPeturbed cell type:\\t\"))\n",
    "        cell_cell_interaction['cell_type_perturbed'] = cell_types[question2_6_5b - 1]\n",
    "        question2_6_5c = int(input(f\"\\tAdjacent cell type:\\t\"))\n",
    "        while question2_6_5c == question2_6_5b:\n",
    "            print(f'\\tAdjacent cell type cannot be the same as peturbed cell type. Please specify another cell type:\\t')\n",
    "            question2_6_5c = int(input(f\"\\tAdjacent cell type:\\t\"))\n",
    "        cell_cell_interaction['cell_type_adj'] = cell_types[question2_6_5c - 1]\n",
    "\n",
    "        question2_6_5d = input(f\"\\tInteraction distance threshold (default = 0.1):\\t\")\n",
    "        if question2_6_5d == '':\n",
    "            question2_6_5d = '0.1'\n",
    "            print('\\tUsing default 0.1.')\n",
    "        cell_cell_interaction['dist_cutoff'] = question2_6_5d\n",
    "\n",
    "        question2_6_5e = input(f\"\\tDo you want to upload gene ID (y/n) [default: n. The proportion of genes with the pattern should be specified, and genes will be randomly selected.]\\t\").lower()\n",
    "        if question2_6_5e == '':\n",
    "            question2_6_5e = 'n'\n",
    "            print('\\tUsing default: n.')\n",
    "        \n",
    "        if question2_6_5e == 'y':\n",
    "            print_workdir_files()\n",
    "            print('\\n')\n",
    "            question2_6_5e_userinput = input(f\"\\tWhich file are you using as gene ID (1-{len(workdir_files)}): \")\n",
    "            geneid = pd.read_csv(workdir_files[int(question2_6_5e_userinput)-1], sep=\"\\t\", header=None)\n",
    "            cell_cell_interaction['gene_id1'] = ','.join(list(geneid[0]))\n",
    "            cell_cell_interaction['gene_prop'] = 'NULL'\n",
    "        else:\n",
    "            cell_cell_interaction['gene_id1'] = \"NULL\"\n",
    "            question2_6_5f = input(f\"\\tGene proportion (default = 0.1):\\t\")\n",
    "            if question2_6_5f == '':\n",
    "                question2_6_5f = '0.1'\n",
    "                print('\\tUsing default 0.1.')\n",
    "            cell_cell_interaction['gene_prop'] = question2_6_5f\n",
    "\n",
    "        question2_6_5g = input(f\"\\tMean effect at log(count) scale (default = 0.5):\")\n",
    "        if question2_6_5g == '':\n",
    "            question2_6_5g = '0.5'\n",
    "            print('\\tUsing default 0.5.')\n",
    "    \n",
    "        cell_cell_interaction['mean'] = question2_6_5g\n",
    "\n",
    "        question2_6_5h = input(f\"\\tSD of effect at log(count) scale (default = 0):\\t\")\n",
    "\n",
    "        if question2_6_5h == '':\n",
    "            question2_6_5h = '0'\n",
    "            print('\\tUsing default 0.')\n",
    "        cell_cell_interaction['sd'] = question2_6_5h\n",
    "\n",
    "        cell_cell_interactions.append(cell_cell_interaction)\n",
    "        print('\\n')\n",
    "        question2_6_5exit = input(f\"\\tSpecify another cell-cell interaction - expression associated with cell-cell distance? (y/n) [default: n]\\t\").lower()\n",
    "        if question2_6_5exit == 'n' or question2_6_5exit == '':\n",
    "            finished_question2_6_5 = True\n",
    "\n",
    "    for i, cell_cell_interaction in enumerate(cell_cell_interactions):\n",
    "        for k in cell_cell_interaction.keys():\n",
    "            parameters[f'spatial_int_dist_{i+1}_{k}'] = cell_cell_interaction[k]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('\\n')\n",
    "question2_6_6 = input(\"\\tAdd cell-cell interactions - expression associated with expression of neighboring cells? (y/n) [default: n]\\t\").lower()\n",
    "\n",
    "if question2_6_6 == 'y':\n",
    "    cell_cell_interactions = []\n",
    "    finished_question2_6_6 = False\n",
    "    while not finished_question2_6_6:\n",
    "        cell_cell_interaction = {}\n",
    "        print('\\n')\n",
    "        print('\\tPrompts for adding cell-cell interaction: ')\n",
    "\n",
    "        if sim_num_regions > 1:\n",
    "            question2_6_6a = input(f\"\\tWhich region? [1-{sim_num_regions}]\\t\").lower()\n",
    "            if not question2_6_6a.isnumeric():\n",
    "                raise ValueError(f\"Please specify a region that is an integer between 1 and {sim_num_regions}\")\n",
    "            cell_cell_interaction['region'] = question2_6_5a\n",
    "        elif cellfeature_num_regions > 1:\n",
    "            print('\\n')\n",
    "            print(f'\\tChoose from regions')\n",
    "            print('\\t--------------')\n",
    "            region_list = list(set(cellfeature_data_r.iloc[:,3]))\n",
    "            for i, region in enumerate(region_list):\n",
    "                print(f'\\t{i+1}: {region}')\n",
    "            question2_6_6a = int(input(f'\\tPlease specify a region:\\t'))\n",
    "            cell_cell_interaction['region'] = region_list[question2_6_6a - 1]\n",
    "        else:\n",
    "            cell_cell_interaction['region'] = 'NULL'\n",
    "        \n",
    "        print('\\n')\n",
    "        print(f'\\tChoose from cell types')\n",
    "        print('\\t--------------')\n",
    "        cell_types = cellfeature_data_cell_types\n",
    "        for i, cell_type in enumerate(cell_types):\n",
    "            print(f'\\t{i+1}: {cell_type}')\n",
    "\n",
    "        question2_6_6b = int(input(f\"\\tPeturbed cell type:\\t\"))\n",
    "        cell_cell_interaction['cell_type_perturbed'] = cell_types[question2_6_6b - 1]\n",
    "        question2_6_6c = int(input(f\"\\tAdjacent cell type:\\t\"))\n",
    "        while question2_6_6c == question2_6_6b:\n",
    "            print(f'\\tAdjacent cell type cannot be the same as peturbed cell type. Please specify another cell type:\\t')\n",
    "            question2_6_5c = int(input(f\"\\tAdjacent cell type:\\t\"))\n",
    "        cell_cell_interaction['cell_type_adj'] = cell_types[question2_6_6c - 1]\n",
    "\n",
    "        question2_6_6d = input(f\"\\tInteraction distance threshold (default = 0.1):\\t\")\n",
    "        if question2_6_6d == '':\n",
    "            question2_6_6d = '0.1'\n",
    "            print('\\tUsing default 0.1.')\n",
    "        cell_cell_interaction['dist_cutoff'] = question2_6_6d\n",
    "\n",
    "        question2_6_6e = input(f\"\\tDo you want to upload gene pair ID (y/n)\\nInstruction: Should be K by 2 gene-pair matrix, where K is the number of gene pairs, and column 2 gives the gene ID in perturbed and adjacent cell types. If no, the proportion of genes with this pattern should be specified, and gene pairs will be randomly generated. [Default: n. The proportion of genes with this pattern should be specified and genes will be randomly selected.]\\t\").lower()\n",
    "        if question2_6_6e == '':\n",
    "            question2_6_6e = 'n'\n",
    "            print('\\tUsing default: n.')\n",
    "\n",
    "        if question2_6_6e == 'y':\n",
    "            print_workdir_files()\n",
    "            print('\\n')\n",
    "            question2_6_6e_userinput = input(f\"\\tWhich file are you using as gene pair ID (1-{len(workdir_files)}): \")\n",
    "            genepair = pd.read_csv(workdir_files[int(question2_6_6e_userinput)-1], sep=\"\\t\", header=None)\n",
    "            cell_cell_interaction['gene_id1'] = ','.join(list(genepair[0]))\n",
    "            cell_cell_interaction['gene_id2'] = ','.join(list(genepair[1]))\n",
    "            cell_cell_interaction['gene_prop'] = 'NULL'\n",
    "        else:\n",
    "            cell_cell_interaction['gene_id1'] = 'NULL'\n",
    "            cell_cell_interaction['gene_id2'] = 'NULL'    \n",
    "            question2_6_6f = input(f\"\\tGene proportion (default = 0.1):\\t\")\n",
    "            if question2_6_6f == '':\n",
    "                question2_6_6f = '0.1'\n",
    "                print('\\tUsing default 0.1.')\n",
    "            cell_cell_interaction['gene_prop'] = question2_6_6f\n",
    "\n",
    "        question2_6_6g = input(f\"\\tIs the association bidirectional (y/n): [default = y]\")\n",
    "        if question2_6_6g == '':\n",
    "            question2_6_6g = 'TRUE'\n",
    "            print('\\tUsing default TRUE')\n",
    "    \n",
    "        cell_cell_interaction['bidirectional'] = question2_6_6g\n",
    "\n",
    "        question2_6_6h = input(f\"\\tMean effect at log(count) scale (default = 0.5):\")\n",
    "        if question2_6_6h == '':\n",
    "            question2_6_6h = '0.5'\n",
    "            print('\\tUsing default 0.5.')\n",
    "    \n",
    "        cell_cell_interaction['mean'] = question2_6_6h\n",
    "\n",
    "        question2_6_6i = input(f\"\\tSD of effect at log(count) scale (default = 0):\\t\")\n",
    "\n",
    "        if question2_6_6i == '':\n",
    "            question2_6_6i = '0'\n",
    "            print('\\tUsing default 0.')\n",
    "        cell_cell_interaction['sd'] = question2_6_6i\n",
    "\n",
    "        cell_cell_interactions.append(cell_cell_interaction)\n",
    "        print('\\n')\n",
    "        question2_6_6exit = input(f\"\\tSpecify another cell-cell interactions - expression associated with expression of neighboring cells? (y/n) [default: n]\\t\").lower()\n",
    "        if question2_6_6exit == 'n' or question2_6_6exit == '':\n",
    "            finished_question2_6_6 = True\n",
    "\n",
    "    for i, cell_cell_interaction in enumerate(cell_cell_interactions):\n",
    "        for k in cell_cell_interaction.keys():\n",
    "            parameters[f'spatial_int_expr_{i+1}_{k}'] = cell_cell_interaction[k]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# question2_12\n",
    "print(\"\\n\")\n",
    "question2_12 = input(\"\\tIf multi-cell resolution data should be simulated, specify the No. of spots [default: NULL = single-cell resolution data is simulated]\\t\")\n",
    "\n",
    "if question2_12 == '':\n",
    "    question2_12 = 'NULL'\n",
    "    print(f\"\\tUsing default: {question2_12}\")\n",
    "\n",
    "parameters['num_spots'] = question2_12"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# question2_7\n",
    "print(\"\\n\")\n",
    "question2_7 = input(\"\\tNumber of simulated data sets [default = 1]: \\t\").lower()\n",
    "\n",
    "if question2_7 == '':\n",
    "    question2_7 = '1'\n",
    "    print(f\"\\tUsing default: {question2_7}\")\n",
    "\n",
    "if not question2_7.isnumeric():\n",
    "    raise ValueError(f'Please enter an integer for the number of simulated data sets. You entered {question2_7}')\n",
    "\n",
    "parameters['num_simulated_datasets'] = question2_7"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question2_8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# question2_8\n",
    "print(\"\\n\")\n",
    "question2_8 = input(\"\\tSpecify an umbrella seed for reproducible simulation [default = 123]: \\t\").lower()\n",
    "\n",
    "if question2_8 == '':\n",
    "    question2_8 = '123'\n",
    "    print(f\"\\tUsing default: {question2_8}\")\n",
    "\n",
    "if not question2_8.isnumeric():\n",
    "    raise ValueError(f'Please enter an integer for the umbrella seed. You entered {question2_8}')\n",
    "\n",
    "parameters['parent_simulation_seed'] = question2_8\n",
    "\n",
    "simulation_seeds = [str(int(parameters['parent_simulation_seed']) + i) for i in range(0, int(parameters['num_simulated_datasets']))]\n",
    "parameters['simulation_seed_for_each_dataset'] = ','.join(simulation_seeds)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Use existing parameter file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question 2_9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('\\n')\n",
    "path_to_output_dir = input(\"\\tPath to save output: [default = working_directory/output_files/]\\t\")\n",
    "if path_to_output_dir == '':\n",
    "    path_to_output_dir = 'working_directory/output_files/'\n",
    "    print('\\tUsing default: ', path_to_output_dir)\n",
    "parameters['path_to_output_dir'] = path_to_output_dir\n",
    "\n",
    "print('\\n')\n",
    "output_name = input(\"\\tName to save output: [default = myfile]\\t\")\n",
    "if output_name == '':\n",
    "    output_name = 'myfile'\n",
    "    print('\\tUsing default: ', output_name)\n",
    "parameters['output_name'] = output_name"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save parameter file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameter_series = pd.Series(parameters, name='value')\n",
    "parameter_series.index.name = 'parameters'\n",
    "parameter_series.reset_index()\n",
    "\n",
    "parameter_file_name = input(\"\\n\\tName this parameter file [default='parameter_file']: \").lower()\n",
    "\n",
    "if parameter_file_name == '':\n",
    "    parameter_file_name = 'parameter_file'\n",
    "    print('\\tUsing default: ', parameter_file_name)\n",
    "\n",
    "pd.DataFrame(parameter_series).to_csv(f'working_directory/{parameter_file_name}.tsv', sep='\\t')\n",
    "\n",
    "print(f'\\n\\tSaved parameter file {parameter_file_name}.tsv for future use to your working directory')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate python script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook docker_prompts.ipynb to script\n",
      "[NbConvertApp] Writing 41435 bytes to docker_prompts.py\n"
     ]
    }
   ],
   "source": [
    " try:\n",
    "     !jupyter nbconvert --to script docker_prompts.ipynb\n",
    " except NameError:\n",
    "     print('\\n')\n",
    "     print('\\tFinished')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "27af9b28b6b180c1dbc04b0edbe8dc028299efbc2ab88247676441af934bb168"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
